File: Advent_of_code/Advent_of_code_04.cpp
#include <iostream> 
#include <fstream>
#include <vector>

using namespace std;

vector<int> text_to_int(string input){
    string str = "";
    vector<int> intervals;

    for (int i = 0; i < input.length(); i++){
        if (input[i] != '-' && input[i] != '\n'){
            str += input[i];
        } else {
            intervals.push_back(stoi(str));
            str = "";
        }
    }    
    intervals.push_back(stoi(str));

    return intervals;
}

vector<int> get_data() {
    vector<int> data;
    string line;
    fstream file("data/04_id_cleaning.txt");

    if (file.is_open())
    {
        for (int line_num = 1; getline(file, line, ',') && line != "\n"; line_num++)
        {
            vector<int> intervals = text_to_int(line);
            // cout << "line= " << line << " endl" << endl;
            for (int i = 0; i < intervals.size(); i++) {
                data.push_back(intervals[i]);
            }
        }
    }
    return data;
}

int fully_contains(int low_l, int high_l, int low_r, int high_r){
    /* Groups data and checks whether a set fully contains another one.
       The variable names indicate which set they belong to, where low_l 
       represents the lowest number on the left set, and so on */

    int counter = 0;
    if (low_l == low_r || high_l == high_r) counter += 1;

    else if (low_l < low_r){
        if (high_l >= high_r) counter += 1;
    } 
    else if (low_l > low_r){
        if (high_l <= high_r) counter += 1;
    }

    return counter;
}

int overlaps(int low_l, int high_l, int low_r, int high_r){
    /* Checks if any of the sets overlap with the other. 
       The variable names follow the structure in "fully_contains" */
    int counter = 0;

    if (low_l == low_r || high_l == high_r) counter += 1;

    else if (low_l < low_r){
        if (high_l >= low_r) counter += 1;
    } 
    else if (low_l > low_r){
        if (low_l <= high_r) counter += 1;
    }

    return counter;
}

int main() {
    vector<int> ids = get_data();
    int fully_contained = 0;
    int overlapping = 0;

    for (int i = 0; i < ids.size(); i+=4){
        fully_contained += fully_contains(ids[i], ids[i+1], ids[i+2], ids[i+3]);
        overlapping += overlaps(ids[i], ids[i+1], ids[i+2], ids[i+3]);
    }

    cout << fully_contained << " pairs fully contain another" << endl;
    cout << overlapping << " pairs overlap" << endl;

    return 1;
}

File: Advent_of_code/Advent_of_code_04.py
def fully_contains(data):
    """Checks how many pairs are contained within another"""
    counter = 0

    for i in data:
        if int(i[0][0]) < int(i[1][0]):
            if int(i[0][1]) >= int(i[1][1]):
                counter+= 1

        elif int(i[1][0]) < int(i[0][0]):
            if int(i[1][1]) >= int(i[0][1]):
                counter += 1

        elif int(i[1][0]) == int(i[0][0]) or int(i[0][1]) == int(i[1][1]):
                counter += 1

    print(f"{counter} pairs fully contain the other")

def overlaps(data):
    """Counts how many pairs have overlaps on their ranges"""
    counter = 0

    for i in data:
        if int(i[0][0]) == int(i[1][0]) or int(i[0][1]) == int(i[1][1]):
            counter += 1

        elif int(i[0][0]) < int(i[1][0]):
            if int(i[0][1]) >= int(i[1][0]):
                counter+= 1

        elif int(i[1][0]) < int(i[0][0]):
            if int(i[1][1]) >= int(i[0][0]):
                counter += 1

    print(f"{counter} pairs overlap")

with open("data/04_id_cleaning.txt", "r") as f:
    file = f.readlines()

# The original data comes in this format: 18-20,19-21
data = [(j[0].split("-"), j[1].split("-")) for j in [i.strip("\n").split(",") for i in file if i != "\n"]]

#After preprocessing, i has the becomes the following: (['18', '20'], ['19', '21']))
fully_contains(data)
overlaps(data)